### PR TITLE
hardening/1359_pre_aggregate_batches_sth

### PR DESCRIPTION
**Pull request by frbattid**
* Implements issue #1359 
* (unofficial) e2e tests passed:
```
time=2017-04-10T08:44:17.605UTC | lvl=INFO | corr=fed779e0-9c88-4861-87f5-fe3fcc3e55cd | trans=fed779e0-9c88-4861-87f5-fe3fcc3e55cd | srv=sc_e2e_test | subsrv=/gardens | comp=cygnus-ngsi | op=getEvents | msg=com.telefonica.iot.cygnus.handlers.NGSIRestHandler[286] : [NGSIRestHandler] Starting internal transaction (fed779e0-9c88-4861-87f5-fe3fcc3e55cd)
time=2017-04-10T08:44:17.606UTC | lvl=INFO | corr=fed779e0-9c88-4861-87f5-fe3fcc3e55cd | trans=fed779e0-9c88-4861-87f5-fe3fcc3e55cd | srv=sc_e2e_test | subsrv=/gardens | comp=cygnus-ngsi | op=getEvents | msg=com.telefonica.iot.cygnus.handlers.NGSIRestHandler[304] : [NGSIRestHandler] Received data ({  "subscriptionId" : "51c0ac9ed714fb3b37d7d5a8",  "originator" : "localhost",  "contextResponses" : [    {      "contextElement" : {        "attributes" : [          {            "name" : "temperature",            "type" : "centigrade",            "value" : "26.5"          },          {            "name" : "pressure",            "type" : "pascal",            "value" : "78.1"          },          {            "name" : "message",            "type" : "string",            "value" : "correct"          },          {            "name" : "status",            "type" : "string",            "value" : "ok"          }        ],        "type" : "Room",        "isPattern" : "false",        "id" : "Room1"      },      "statusCode" : {        "code" : "200",        "reasonPhrase" : "OK"      }    }  ]})
time=2017-04-10T08:44:19.259UTC | lvl=INFO | corr=14a09aa2-20b4-4832-a41b-d02e0558bf67 | trans=14a09aa2-20b4-4832-a41b-d02e0558bf67 | srv=sc_e2e_test | subsrv=/gardens | comp=cygnus-ngsi | op=getEvents | msg=com.telefonica.iot.cygnus.handlers.NGSIRestHandler[286] : [NGSIRestHandler] Starting internal transaction (14a09aa2-20b4-4832-a41b-d02e0558bf67)
time=2017-04-10T08:44:19.259UTC | lvl=INFO | corr=14a09aa2-20b4-4832-a41b-d02e0558bf67 | trans=14a09aa2-20b4-4832-a41b-d02e0558bf67 | srv=sc_e2e_test | subsrv=/gardens | comp=cygnus-ngsi | op=getEvents | msg=com.telefonica.iot.cygnus.handlers.NGSIRestHandler[304] : [NGSIRestHandler] Received data ({  "subscriptionId" : "51c0ac9ed714fb3b37d7d5a8",  "originator" : "localhost",  "contextResponses" : [    {      "contextElement" : {        "attributes" : [          {            "name" : "temperature",            "type" : "centigrade",            "value" : "26.5"          },          {            "name" : "pressure",            "type" : "pascal",            "value" : "78.1"          },          {            "name" : "message",            "type" : "string",            "value" : "correct"          },          {            "name" : "status",            "type" : "string",            "value" : "ok"          }        ],        "type" : "Room",        "isPattern" : "false",        "id" : "Room1"      },      "statusCode" : {        "code" : "200",        "reasonPhrase" : "OK"      }    }  ]})
time=2017-04-10T08:44:20.057UTC | lvl=INFO | corr=044f24b9-f6d5-4bf5-8764-c99d3cf15c44 | trans=044f24b9-f6d5-4bf5-8764-c99d3cf15c44 | srv=sc_e2e_test | subsrv=/gardens | comp=cygnus-ngsi | op=getEvents | msg=com.telefonica.iot.cygnus.handlers.NGSIRestHandler[286] : [NGSIRestHandler] Starting internal transaction (044f24b9-f6d5-4bf5-8764-c99d3cf15c44)
time=2017-04-10T08:44:20.058UTC | lvl=INFO | corr=044f24b9-f6d5-4bf5-8764-c99d3cf15c44 | trans=044f24b9-f6d5-4bf5-8764-c99d3cf15c44 | srv=sc_e2e_test | subsrv=/gardens | comp=cygnus-ngsi | op=getEvents | msg=com.telefonica.iot.cygnus.handlers.NGSIRestHandler[304] : [NGSIRestHandler] Received data ({  "subscriptionId" : "51c0ac9ed714fb3b37d7d5a8",  "originator" : "localhost",  "contextResponses" : [    {      "contextElement" : {        "attributes" : [          {            "name" : "temperature",            "type" : "centigrade",            "value" : "26.5"          },          {            "name" : "pressure",            "type" : "pascal",            "value" : "78.1"          },          {            "name" : "message",            "type" : "string",            "value" : "correct"          },          {            "name" : "status",            "type" : "string",            "value" : "ok"          }        ],        "type" : "Room",        "isPattern" : "false",        "id" : "Room1"      },      "statusCode" : {        "code" : "200",        "reasonPhrase" : "OK"      }    }  ]})
time=2017-04-10T08:44:51.879UTC | lvl=INFO | corr=044f24b9-f6d5-4bf5-8764-c99d3cf15c44 | trans=044f24b9-f6d5-4bf5-8764-c99d3cf15c44 | srv=sc_e2e_test | subsrv=/gardens | comp=cygnus-ngsi | op=persist | msg=com.telefonica.iot.cygnus.sinks.NGSISTHSink$STHAggregator[380] : [sth-sink] Persisting data at NGSISTHSink. Database: sth_sc_e2e_test, Collection: sth_/gardens_Room1_Room.aggr, Data: 1491813860059,Room1,Room,pressure,pascal,[78.1,78.1,234.29999999999998,18298.829999999994,3]
time=2017-04-10T08:44:52.357UTC | lvl=INFO | corr=044f24b9-f6d5-4bf5-8764-c99d3cf15c44 | trans=044f24b9-f6d5-4bf5-8764-c99d3cf15c44 | srv=sc_e2e_test | subsrv=/gardens | comp=cygnus-ngsi | op=persist | msg=com.telefonica.iot.cygnus.sinks.NGSISTHSink$STHAggregator[380] : [sth-sink] Persisting data at NGSISTHSink. Database: sth_sc_e2e_test, Collection: sth_/gardens_Room1_Room.aggr, Data: 1491813860059,Room1,Room,temperature,centigrade,[26.5,26.5,79.5,2106.75,3]
time=2017-04-10T08:44:52.597UTC | lvl=INFO | corr=044f24b9-f6d5-4bf5-8764-c99d3cf15c44 | trans=044f24b9-f6d5-4bf5-8764-c99d3cf15c44 | srv=sc_e2e_test | subsrv=/gardens | comp=cygnus-ngsi | op=persist | msg=com.telefonica.iot.cygnus.sinks.NGSISTHSink$STHAggregator[400] : [sth-sink] Persisting data at NGSISTHSink. Database: sth_sc_e2e_test, Collection: sth_/gardens_Room1_Room.aggr, Data: 1491813860059,Room1,Room,message,string,[{correct=3},3]
time=2017-04-10T08:44:52.758UTC | lvl=INFO | corr=044f24b9-f6d5-4bf5-8764-c99d3cf15c44 | trans=044f24b9-f6d5-4bf5-8764-c99d3cf15c44 | srv=sc_e2e_test | subsrv=/gardens | comp=cygnus-ngsi | op=persist | msg=com.telefonica.iot.cygnus.sinks.NGSISTHSink$STHAggregator[400] : [sth-sink] Persisting data at NGSISTHSink. Database: sth_sc_e2e_test, Collection: sth_/gardens_Room1_Room.aggr, Data: 1491813860059,Room1,Room,status,string,[{ok=3},3]
time=2017-04-10T08:44:52.915UTC | lvl=INFO | corr=044f24b9-f6d5-4bf5-8764-c99d3cf15c44 | trans=044f24b9-f6d5-4bf5-8764-c99d3cf15c44 | srv=sc_e2e_test | subsrv=/gardens | comp=cygnus-ngsi | op=processNewBatches | msg=com.telefonica.iot.cygnus.sinks.NGSISink[553] : Finishing internal transaction (fed779e0-9c88-4861-87f5-fe3fcc3e55cd,14a09aa2-20b4-4832-a41b-d02e0558bf67,044f24b9-f6d5-4bf5-8764-c99d3cf15c44)
..
..
..
{ "_id" : { "attrName" : "pressure", "origin" : ISODate("2017-04-10T08:00:00Z"), "resolution" : "minute", "range" : "hour" }, "points" : [...{ "offset" : 44, "samples" : 3, "sum" : 234.29999999999998, "sum2" : 18298.829999999994, "min" : 78.1, "max" : 78.1 }...] }, "attrType" : "pascal" }
{ "_id" : { "attrName" : "pressure", "origin" : ISODate("2017-04-10T00:00:00Z"), "resolution" : "hour", "range" : "day" }, "points" : [...{ "offset" : 8, "samples" : 3, "sum" : 234.29999999999998, "sum2" : 18298.829999999994, "min" : 78.1, "max" : 78.1 }...] }, "attrType" : "pascal" }
{ "_id" : { "attrName" : "pressure", "origin" : ISODate("2017-04-01T00:00:00Z"), "resolution" : "day", "range" : "month" }, "points" : [...{ "offset" : 10, "samples" : 3, "sum" : 234.29999999999998, "sum2" : 18298.829999999994, "min" : 78.1, "max" : 78.1 }...} ], "attrType" : "pascal" }
{ "_id" : { "attrName" : "pressure", "origin" : ISODate("2017-01-01T00:00:00Z"), "resolution" : "month", "range" : "year" }, "points" : [...{ "offset" : 4, "samples" : 3, "sum" : 234.29999999999998, "sum2" : 18298.829999999994, "min" : 78.1, "max" : 78.1 }...} ], "attrType" : "pascal" }
{ "_id" : { "attrName" : "temperature", "origin" : ISODate("2017-04-10T08:00:00Z"), "resolution" : "minute", "range" : "hour" }, "points" : [...{ "offset" : 44, "samples" : 3, "sum" : 79.5, "sum2" : 2106.75, "min" : 26.5, "max" : 26.5 }...} ], "attrType" : "centigrade" }
{ "_id" : { "attrName" : "temperature", "origin" : ISODate("2017-04-10T00:00:00Z"), "resolution" : "hour", "range" : "day" }, "points" : [...{ "offset" : 8, "samples" : 3, "sum" : 79.5, "sum2" : 2106.75, "min" : 26.5, "max" : 26.5 }...} ], "attrType" : "centigrade" }
{ "_id" : { "attrName" : "temperature", "origin" : ISODate("2017-04-01T00:00:00Z"), "resolution" : "day", "range" : "month" }, "points" : [...{ "offset" : 10, "samples" : 3, "sum" : 79.5, "sum2" : 2106.75, "min" : 26.5, "max" : 26.5 }...} ], "attrType" : "centigrade" }
{ "_id" : { "attrName" : "temperature", "origin" : ISODate("2017-01-01T00:00:00Z"), "resolution" : "month", "range" : "year" }, "points" : [...{ "offset" : 4, "samples" : 3, "sum" : 79.5, "sum2" : 2106.75, "min" : 26.5, "max" : 26.5 }...} ], "attrType" : "centigrade" }
{ "_id" : { "attrName" : "message", "origin" : ISODate("2017-04-10T08:00:00Z"), "resolution" : "minute", "range" : "hour" }, "points" : [...{ "offset" : 44, "samples" : 3, "occur" : { "correct" : 3 } }...} ], "attrType" : "string" }
{ "_id" : { "attrName" : "message", "origin" : ISODate("2017-04-10T00:00:00Z"), "resolution" : "hour", "range" : "day" }, "points" : [...{ "offset" : 8, "samples" : 3, "occur" : { "correct" : 3 } }...} ], "attrType" : "string" }
{ "_id" : { "attrName" : "message", "origin" : ISODate("2017-04-01T00:00:00Z"), "resolution" : "day", "range" : "month" }, "points" : [...{ "offset" : 10, "samples" : 3, "occur" : { "correct" : 3 } }...} ], "attrType" : "string" }
{ "_id" : { "attrName" : "message", "origin" : ISODate("2017-01-01T00:00:00Z"), "resolution" : "month", "range" : "year" }, "points" : [...{ "offset" : 4, "samples" : 3, "occur" : { "correct" : 3 } }...} ], "attrType" : "string" }
{ "_id" : { "attrName" : "status", "origin" : ISODate("2017-04-10T08:00:00Z"), "resolution" : "minute", "range" : "hour" }, "points" : [...{ "offset" : 44, "samples" : 3, "occur" : { "ok" : 3 } }...} ], "attrType" : "string" }
{ "_id" : { "attrName" : "status", "origin" : ISODate("2017-04-10T00:00:00Z"), "resolution" : "hour", "range" : "day" }, "points" : [...{ "offset" : 8, "samples" : 3, "occur" : { "ok" : 3 } }...} ], "attrType" : "string" }
{ "_id" : { "attrName" : "status", "origin" : ISODate("2017-04-01T00:00:00Z"), "resolution" : "day", "range" : "month" }, "points" : [...{ "offset" : 10, "samples" : 3, "occur" : { "ok" : 3 } }...} ], "attrType" : "string" }
{ "_id" : { "attrName" : "status", "origin" : ISODate("2017-01-01T00:00:00Z"), "resolution" : "month", "range" : "year" }, "points" : [...{ "offset" : 4, "samples" : 3, "occur" : { "ok" : 3 } }...} ], "attrType" : "string" }
```

### CODE REVIEW

File: cygnus-common/src/main/java/com/telefonica/iot/cygnus/backends/mongo/MongoBackend.java [hunk 1]
```diff
@@ -66,42 +67,45 @@ void createCollection(String dbName, String collectionName, long collectionsSize
      * @throws Exception
      */
     void insertContextDataRaw(String dbName, String collectionName, ArrayList<Document> aggregation) throws Exception;
-
+    
     /**
-     * Inserts a new document in the given aggregated collection within the given database (row-like mode).
+     * Inserts pre-aggregated numeric context data in the related aggregated collction for the resolutions given.
```
Comment (gtorodelvalle) [inline cygnus-common/src/main/java/com/telefonica/iot/cygnus/backends/mongo/MongoBackend.java]: Minor typo: collction -> collection

File: cygnus-common/src/main/java/com/telefonica/iot/cygnus/backends/mongo/MongoBackend.java [hunk 2]
```diff
@@ -66,42 +67,45 @@ void createCollection(String dbName, String collectionName, long collectionsSize
...
+            String entityType, String attrName, String attrType, double max, double min, double sum, double sum2,
+            int numSamples, boolean[] resolutions)
         throws Exception;
-
+    
     /**
-     * Stores in per-service/database "collection_names" collection the matching between a hash and the fields used to
-     * build it.
-     * FIXME: destination is under study
+     * Inserts pre-aggregated string-based context data in the related aggregated collction for the resolutions given.
```
Comment (gtorodelvalle) [inline cygnus-common/src/main/java/com/telefonica/iot/cygnus/backends/mongo/MongoBackend.java]: Minor typo: collction -> collection

File: cygnus-common/src/main/java/com/telefonica/iot/cygnus/backends/mongo/MongoBackend.java (hunk 1 above)
Comment (frbattid) [inline cygnus-common/src/main/java/com/telefonica/iot/cygnus/backends/mongo/MongoBackend.java]: Fixed in https://github.com/telefonicaid/fiware-cygnus/pull/1426/commits/a3bb342105b0c86818d833b69735985f450f02f8

File: cygnus-common/src/main/java/com/telefonica/iot/cygnus/backends/mongo/MongoBackend.java (hunk 2 above)
Comment (frbattid) [inline cygnus-common/src/main/java/com/telefonica/iot/cygnus/backends/mongo/MongoBackend.java]: Fixed in https://github.com/telefonicaid/fiware-cygnus/pull/1426/commits/a3bb342105b0c86818d833b69735985f450f02f8

**Pull request merged**